src/ctap/crypto_wrapper.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::ctap::status_code::Ctap2StatusCode;
+use alloc::vec;
+use alloc::vec::Vec;
+use crypto::cbc::{cbc_decrypt, cbc_encrypt};
+use crypto::rng256::Rng256;
+
+/// Wraps the AES256-CBC encryption to match what we need in CTAP.
+pub fn aes256_cbc_encrypt(
+    rng: &mut dyn Rng256,
+    aes_enc_key: &crypto::aes256::EncryptionKey,
+    plaintext: &[u8],
+    has_iv: bool,
```

Comment:
CBC always need an IV. So this boolean is actually NULL IV vs. random IV.
Why don't we create the IV in the caller and pass it as an arg? The API would be easier to understand.
It would also make the API more consistent with `aes256_cbc_decrypt`.

That being said, when we'll move to protected implementations, this abstraction will likely change.

Reply:
The output with `has_iv = false` is not equivalent to passing in a zero IV. We only prepend the IV if we don't use the implicit zeros. The caller could of course drop the unused IV from the result, the required changes are:
- For `has_iv = false` use cases, pass in a zero IV and remove it from the output.
- For `has_iv = true` use cases, generate the random IV and move it into the function.

This makes it easier to get function calls wrong IMO. I agree that the API is not great for general crypto. It's supposed to be tailored for CTAP use cases. WDYT?

Reply:
Ok, I was missing this part indeed.
What about `embeds_iv` or something alike?

Reply:
I can rename it!
We could simplify the interface and remove the RNG from the parameter list by using an `iv: Option<[u8; 16]>`. It still means the caller has to correctly add entropy themselves, but it would make the encrypt deterministic. On the other hand, the decrypt still only needs a `bool` to know whether the IV is part of the input `Vec` or not, which makes it a bit inconsistent.

Reply:
Renaming done. I currently prefer the `bool` over the `Option`, but let me know if you prefer the latter.